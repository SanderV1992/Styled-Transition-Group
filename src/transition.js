import React from 'react'
import PropTypes from 'prop-types'

import styled, { css } from 'styled-components'
import { TransitionGroup as ReactTransitionGroup, CSSTransition as ReactCSSTransition } from 'react-transition-group'


const defaultProps = {
  type: 'fade',
  duration: 1000,
}

const propTypes = {
  className: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  type: PropTypes.string,
  duration: PropTypes.number,
}

export const transitionStyles = {
  fade: css`
    &.fade-enter {
      opacity: 0.01;
    }
    
    &.fade-enter.fade-enter-active {
      opacity: 1;
    }
    
    &.fade-exit {
      opacity: 1;
    }
    
    &.fade-exit.fade-exit-active {
      opacity: 0.01;
    }
  `,
  zoom: css`
   &.zoom-enter {
      opacity: 0.01;
      transform: scale3d(0.3, 0.3, 0.3);
    }
    
    &.zoom-enter.zoom-enter-active {
      opacity: 1;
      transform: initial;
    }
    
    &.zoom-exit {
      opacity: 1;
      transform: initial;
    }
    
    &.zoom-exit.zoom-exit-active {
      opacity: 0.01;
      transform: scale3d(0.3, 0.3, 0.3);
    }
  `,
  rotate: css`
    &.rotate-enter {
      opacity: 0.01;
      transform: rotate3d(0, 0, 1, -200deg);
    }
    
    &.rotate-enter.rotate-enter-active {
      opacity: 1;
      transform: initial;
    }
    
    &.rotate-exit {
      opacity: 1;
      transform: initial;
    }
    
    &.rotate-exit.rotate-exit-active {
      opacity: 0.01;
      transform: rotate3d(0, 0, 1, -200deg);
    }
  `,
  roll: css`
    &.roll-enter {
      opacity: 0.01;
      transform: translate3d(100%, 0, 0) rotate3d(0, 0, 1, 120deg);
    }
    
    &.roll-enter.roll-enter-active {
      opacity: 1;
      transform: initial;
    }
    
    &.roll-exit {
      opacity: 1;
      transform: initial;
    }
    
    &.roll-exit.roll-exit-active {
      opacity: 0.01;
      transform: translate3d(100%, 0, 0) rotate3d(0, 0, 1, 120deg);
    }
  `,
}

const TransitionConstructor = ({ children, className, type, duration, ...props }) => (
  <ReactCSSTransition
    {...props}
    timeout={duration}
    classNames={type}
    className={className}
  >
    {children}
  </ReactCSSTransition>
)

TransitionConstructor.defaultProps = defaultProps
TransitionConstructor.propTypes = propTypes

export const TransitionGroup = ReactTransitionGroup

export const Transition = styled(TransitionConstructor)`
  transition: all ${(p) => p.duration}ms linear;

  ${(p) => transitionStyles[p.type] && css`
    ${transitionStyles[p.type]}
  `}
`

Transition.defaultProps = defaultProps
