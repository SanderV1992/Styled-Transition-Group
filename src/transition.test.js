import React from 'react'
import ReactDOM from 'react-dom'
import { css } from 'styled-components'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import { TransitionGroup, Transition } from './transition'


describe('TransitionGroup && Transition Components', () => {
  const div = document.createElement('div')
  const animations = ['fade', 'zoom', 'rotate', 'roll']

  let Component
  let ComponentDefault
  let ComponentAnimation

  const animation = css`
      @keyframes bounceInDown {
        from,
        60%,
        75%,
        90%,
        to {
          animation-timing-function: cubic-bezier(0.215, 0.61, 0.355, 1);
        }
  
        0% {
          opacity: 0;
          transform: translate3d(0, -3000px, 0);
        }
  
        60% {
          opacity: 1;
          transform: translate3d(0, 25px, 0);
        }
  
        75% {
          transform: translate3d(0, -10px, 0);
        }
  
        90% {
          transform: translate3d(0, 5px, 0);
        }
  
        to {
          transform: translate3d(0, 0, 0);
        }
      }
  
      @keyframes bounceOutDown {
        20% {
          transform: translate3d(0, 10px, 0);
        }
  
        40%,
        45% {
          opacity: 1;
          transform: translate3d(0, -20px, 0);
        }
  
        to {
          opacity: 0;
          transform: translate3d(0, 2000px, 0);
        }
      }
  
     &.bounce-enter {
        animation: bounceOutDown ${(p) => p.duration}ms linear infinite
      }
  
      &.bounce-enter.bounce-enter-active {
        animation: bounceInDown ${(p) => p.duration}ms linear infinite;
      }
  
      &.bounce-exit {
        animation: bounceInDown ${(p) => p.duration}ms linear infinite;
      }
  
      &.bounce-exit.bounce-exit-active {
        animation: bounceOutDown ${(p) => p.duration}ms linear infinite
      }
    `

  beforeEach(() => {
    Component = () => (
      <TransitionGroup>
        {animations.map((item, index) => (
          <Transition key={item} type={item} duration={2000}>
            <div>
              {index}
            </div>
          </Transition>
        ))}
      </TransitionGroup>
    )

    ComponentDefault = () => (
      <TransitionGroup>
        {animations.map((item, index) => (
          <Transition key={item}>
            <div>
              {index}
            </div>
          </Transition>
        ))}
      </TransitionGroup>
    )

    ComponentAnimation = () => (
      <TransitionGroup>
        {animations.map((item, index) => (
          <Transition key={item} animation={animation}>
            <div>
              {index}
            </div>
          </Transition>
        ))}
      </TransitionGroup>
    )
  })

  test('Will not crash', () => {
    ReactDOM.render(
      <Component />
      , div,
    )
  })

  test('Check Default props', () => {
    const component = mount(ComponentDefault())
    const duration = 1000
    const type = 'fade'

    for (let index = 0; index < animations.length; index++) {
      expect(component.props().children[index].props.duration).toEqual(duration)
      expect(component.props().children[index].props.type).toEqual(type)
    }
  })

  test('Count', () => {
    const component = mount(Component())
    const equalCount = 4

    expect(component.props().children.length).toEqual(equalCount)
  })

  test('Duration', () => {
    const component = mount(Component())
    const duration = 2000

    for (let index = 0; index < animations.length; index++) {
      expect(component.props().children[index].props.duration).toEqual(duration)
    }
  })

  test('Class', () => {
    const component = mount(Component())

    for (let index = 0; index < animations.length; index++) {
      expect(component.props().children[index].type.styledComponentId).toBeDefined()
    }
  })

  test('Children', () => {
    const component = mount(Component())

    for (let index = 0; index < animations.length; index++) {
      expect(component.props().children[index].props.children).toBeDefined()
    }
  })

  test('Render default component snapshot', () => {
    const tree = renderer.create(<ComponentDefault />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('Render custom component snapshot', () => {
    const tree = renderer.create(<Component />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('Add custom animation', () => {
    const tree = renderer.create(<ComponentAnimation />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
