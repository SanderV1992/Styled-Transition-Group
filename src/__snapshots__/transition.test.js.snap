// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransitionGroup && Transition Components Add custom animation 1`] = `
.c0 {
  -webkit-transition: all 1000ms linear;
  transition: all 1000ms linear;
}

.c0.fade-enter {
  opacity: 0.01;
}

.c0.fade-enter.fade-enter-active {
  opacity: 1;
}

.c0.fade-exit {
  opacity: 1;
}

.c0.fade-exit.fade-exit-active {
  opacity: 0.01;
}

.c0.bounce-enter {
  -webkit-animation: bounceOutDown 1000ms linear infinite;
  animation: bounceOutDown 1000ms linear infinite;
}

.c0.bounce-enter.bounce-enter-active {
  -webkit-animation: bounceInDown 1000ms linear infinite;
  animation: bounceInDown 1000ms linear infinite;
}

.c0.bounce-exit {
  -webkit-animation: bounceInDown 1000ms linear infinite;
  animation: bounceInDown 1000ms linear infinite;
}

.c0.bounce-exit.bounce-exit-active {
  -webkit-animation: bounceOutDown 1000ms linear infinite;
  animation: bounceOutDown 1000ms linear infinite;
}

<div>
  <div
    className="c0"
  >
    0
  </div>
  <div
    className="c0"
  >
    1
  </div>
  <div
    className="c0"
  >
    2
  </div>
  <div
    className="c0"
  >
    3
  </div>
</div>
`;

exports[`TransitionGroup && Transition Components Render custom component snapshot 1`] = `
.c0 {
  -webkit-transition: all 2000ms linear;
  transition: all 2000ms linear;
}

.c0.fade-enter {
  opacity: 0.01;
}

.c0.fade-enter.fade-enter-active {
  opacity: 1;
}

.c0.fade-exit {
  opacity: 1;
}

.c0.fade-exit.fade-exit-active {
  opacity: 0.01;
}

.c1 {
  -webkit-transition: all 2000ms linear;
  transition: all 2000ms linear;
}

.c1.zoom-enter {
  opacity: 0.01;
  -webkit-transform: scale3d(0.3,0.3,0.3);
  -ms-transform: scale3d(0.3,0.3,0.3);
  transform: scale3d(0.3,0.3,0.3);
}

.c1.zoom-enter.zoom-enter-active {
  opacity: 1;
  -webkit-transform: initial;
  -ms-transform: initial;
  transform: initial;
}

.c1.zoom-exit {
  opacity: 1;
  -webkit-transform: initial;
  -ms-transform: initial;
  transform: initial;
}

.c1.zoom-exit.zoom-exit-active {
  opacity: 0.01;
  -webkit-transform: scale3d(0.3,0.3,0.3);
  -ms-transform: scale3d(0.3,0.3,0.3);
  transform: scale3d(0.3,0.3,0.3);
}

.c2 {
  -webkit-transition: all 2000ms linear;
  transition: all 2000ms linear;
}

.c2.rotate-enter {
  opacity: 0.01;
  -webkit-transform: rotate3d(0,0,1,-200deg);
  -ms-transform: rotate3d(0,0,1,-200deg);
  transform: rotate3d(0,0,1,-200deg);
}

.c2.rotate-enter.rotate-enter-active {
  opacity: 1;
  -webkit-transform: initial;
  -ms-transform: initial;
  transform: initial;
}

.c2.rotate-exit {
  opacity: 1;
  -webkit-transform: initial;
  -ms-transform: initial;
  transform: initial;
}

.c2.rotate-exit.rotate-exit-active {
  opacity: 0.01;
  -webkit-transform: rotate3d(0,0,1,-200deg);
  -ms-transform: rotate3d(0,0,1,-200deg);
  transform: rotate3d(0,0,1,-200deg);
}

.c3 {
  -webkit-transition: all 2000ms linear;
  transition: all 2000ms linear;
}

.c3.roll-enter {
  opacity: 0.01;
  -webkit-transform: translate3d(100%,0,0) rotate3d(0,0,1,120deg);
  -ms-transform: translate3d(100%,0,0) rotate3d(0,0,1,120deg);
  transform: translate3d(100%,0,0) rotate3d(0,0,1,120deg);
}

.c3.roll-enter.roll-enter-active {
  opacity: 1;
  -webkit-transform: initial;
  -ms-transform: initial;
  transform: initial;
}

.c3.roll-exit {
  opacity: 1;
  -webkit-transform: initial;
  -ms-transform: initial;
  transform: initial;
}

.c3.roll-exit.roll-exit-active {
  opacity: 0.01;
  -webkit-transform: translate3d(100%,0,0) rotate3d(0,0,1,120deg);
  -ms-transform: translate3d(100%,0,0) rotate3d(0,0,1,120deg);
  transform: translate3d(100%,0,0) rotate3d(0,0,1,120deg);
}

<div>
  <div
    className="c0"
  >
    0
  </div>
  <div
    className="c1"
  >
    1
  </div>
  <div
    className="c2"
  >
    2
  </div>
  <div
    className="c3"
  >
    3
  </div>
</div>
`;

exports[`TransitionGroup && Transition Components Render default component snapshot 1`] = `
.c0 {
  -webkit-transition: all 1000ms linear;
  transition: all 1000ms linear;
}

.c0.fade-enter {
  opacity: 0.01;
}

.c0.fade-enter.fade-enter-active {
  opacity: 1;
}

.c0.fade-exit {
  opacity: 1;
}

.c0.fade-exit.fade-exit-active {
  opacity: 0.01;
}

<div>
  <div
    className="c0"
  >
    0
  </div>
  <div
    className="c0"
  >
    1
  </div>
  <div
    className="c0"
  >
    2
  </div>
  <div
    className="c0"
  >
    3
  </div>
</div>
`;
